### PR TITLE
[Fix] Multiple image conditionings in a single batch for `StableDiffusionControlNetPipeline`

### PR DESCRIPTION
**Pull request by celestialphineas**
# What does this PR do?

Allow passing multiple image conditionings in a single batch for `StableDiffusionControlNetPipeline`. This feature was mentioned in the comments, but not implemented. 

See #6286 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@patrickvonplaten @sayakpaul

### CODE REVIEW

**Comment by celestialphineas** [comment]
A working example is as below:

```python
controlnet_generated_imgs = double_control_pipe(
  prompt=['girl', 'boy'],
  image=[list(t) for t in zip(canny_imgs, tile_imgs)],
  guidance_scale=4,
  num_inference_steps=20,
  controlnet_conditioning_scale=[ 1.0, 1.0 ],
  control_guidance_start=[ 0.0, 0.0 ],
  control_guidance_end=[ 0.4, 0.5 ],
  num_images_per_prompt=4
).images
```

### `canny_imgs`

image

### `tile_imgs`

image

### Generated images

image


For single ControlNet pipelines, where a list of images is interpreted as different conditionings for a list of prompts, nested image lists are proposed as the input of multiple ControlNet image conditioning in a single batch:

```python
controlnet_generated_imgs = single_control_pipe(
  prompt=['girl', 'boy'], 
  image=[[canny_imgs[0]], [canny_imgs[1]]],
  # image=[canny_imgs[0], tile_imgs[0]],
  guidance_scale=4,
  num_inference_steps=20,
  controlnet_conditioning_scale=1.0,
  control_guidance_start=0.0,
  control_guidance_end=0.5,
  num_images_per_prompt=2
).images
```

This will trigger warnings:

```
You have passed a nested list for `image` with a single controlnet. This will be interpreted as multiple image conditioning.
You have passed 2 prompts. The conditionings will be fixed across the prompts.
```

And will give the result as below:

image

**Comment by sayakpaul** [comment]
Very nice work. Could you maybe add an end-to-end working code to the PR?

File: src/diffusers/pipelines/controlnet/pipeline_controlnet.py
Comment (sayakpaul) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet.py:627]: This should go to `check_image`.

**Comment by celestialphineas** [comment]
Currently, this PR does not meet `utils/check_copies.py` for the `prepare_image` function. I haven't worked on the fix for other ControlNet pipelines yet, but I believe it is safe to fix-copies, as the behavior hasn't been changed for non-list `image` passed to the function.

**Review by sayakpaul — COMMENTED** [review]
Let's also add a fast test for this. 

Very nice work.

File: src/diffusers/pipelines/controlnet/pipeline_controlnet.py
Comment (celestialphineas) [reply src/diffusers/pipelines/controlnet/pipeline_controlnet.py]: I refactored `check_inputs` and `check_image` in my latest commit.

**Comment by celestialphineas** [comment]
A full working example:

```python
import numpy as np
import skimage, torch
from PIL import Image
from diffusers import ControlNetModel, StableDiffusionPipeline, StableDiffusionControlNetPipeline


# Loat models
sd_pipe = StableDiffusionPipeline.from_pretrained('runwayml/stable-diffusion-v1-5', torch_dtype=torch.float16).to('cuda')
canny_controlnet = ControlNetModel.from_pretrained('lllyasviel/control_v11p_sd15_canny/').to('cuda', dtype=sd_pipe.dtype)
tile_controlnet = ControlNetModel.from_pretrained('lllyasviel/control_v11f1e_sd15_tile/').to('cuda', dtype=sd_pipe.dtype)


# Single ControlNet pipeline
single_control_pipe = StableDiffusionControlNetPipeline(
  vae=sd_pipe.vae,
  text_encoder=sd_pipe.text_encoder,
  tokenizer=sd_pipe.tokenizer,
  unet=sd_pipe.unet,
  scheduler=sd_pipe.scheduler,
  safety_checker=sd_pipe.safety_checker,
  feature_extractor=sd_pipe.feature_extractor,
  controlnet=canny_controlnet
)

# Multiple ControlNet pipeline
double_control_pipe = StableDiffusionControlNetPipeline(
  vae=sd_pipe.vae,
  text_encoder=sd_pipe.text_encoder,
  tokenizer=sd_pipe.tokenizer,
  unet=sd_pipe.unet,
  scheduler=sd_pipe.scheduler,
  safety_checker=sd_pipe.safety_checker,
  feature_extractor=sd_pipe.feature_extractor,
  controlnet=[canny_controlnet, tile_controlnet]
)


# Generate test images with sd_pipe
generator_seed = 114514
sd_generated_imgs = sd_pipe(
  prompt='a lovely cat',
  num_inference_steps=20,
  num_images_per_prompt=4,
  generator=torch.Generator(device='cpu').manual_seed(generator_seed)
).images

# These images are used for canny control
canny_imgs = [Image.fromarray(skimage.feature.canny(np.array(img.convert('L')))) for img in sd_generated_imgs]

# These images are used for tile control, they are colored round disks on white backround
def circle_img(color_hex):
  disk_img = np.zeros((512, 512, 3), dtype=np.uint8) + 255
  rr, cc = skimage.draw.disk((256, 256), radius=200)
  # Convert color_hex to RGB array
  disk_img[rr, cc] = np.array([int(color_hex[i:i+2], 16) for i in (1, 3, 5)], dtype=np.uint8)
  return disk_img
tile_imgs = [ 
  Image.fromarray(circle_img(color))
  for color in [ '#89ED87', '#AE45E2', '#E2B845', '#E24545' ]
]

# Example generation for single_control_pipe
single_controlnet_imgs = single_control_pipe(
  prompt=['cat', 'dog'], 
  # Use this nested image configuration only when num_images_per_prompt=2
  # image=[[ canny_imgs[0] ], [ canny_imgs[1] ]],
  # This works for conditioning each prompt with the same images for ControlNet
  image=[ canny_imgs[0], canny_imgs[1] ],
  guidance_scale=4,
  num_inference_steps=20,
  controlnet_conditioning_scale=1.0,
  control_guidance_start=0.0,
  control_guidance_end=0.5,
  # guess_mode=True,
  generator=torch.Generator(device='cpu').manual_seed(generator_seed),
  num_images_per_prompt=2
).images

# Example generation for double_control_pipe
double_controlnet_imgs = double_control_pipe(
  prompt=[ 'cat', 'dog' ],
  # Use this nested image configuration only when num_images_per_prompt=2
  # image=[
  #   [ canny_imgs[0], tile_imgs[0] ],
  #   [ canny_imgs[1], tile_imgs[1] ]
  # ],
  image=[ list(t) for t in zip(canny_imgs, tile_imgs) ],
  # This works for conditioning each prompt with the same images for multi-ControlNet
  # image=[ canny_imgs[0], tile_imgs[0] ],
  guidance_scale=4,
  num_inference_steps=20,
  controlnet_conditioning_scale=[1.0, 1.0],
  control_guidance_start=[0.0, 0.0],
  control_guidance_end=[0.4, 0.5],
  guess_mode=False,
  generator=torch.Generator(device='cpu').manual_seed(generator_seed),
  num_images_per_prompt=4
).images

skimage.io.imshow(np.concatenate(double_controlnet_imgs, axis=1))
skimage.io.show()
```

I'm working on my local model for the code above, so here's what I've got. Enjoy. :)
image

**Comment by celestialphineas** [comment]
> Let's also add a fast test for this.
> 
> Very nice work.

Is it alright to add the test to `tests/pipelines/controlnet/test_controlnet.py`, with functions added to `ControlNetPipelineFastTests` and `StableDiffusionMultiControlNetPipelineFastTests`?

**Comment by sayakpaul** [comment]
before adding tests, let's wait for a review from @yiyixuxu.

**Comment by patrickvonplaten** [comment]
@yiyixuxu could you take a look here?

> > Let's also add a fast test for this.
> > Very nice work.
> 
> Is it alright to add the test to `tests/pipelines/controlnet/test_controlnet.py`, with functions added to `ControlNetPipelineFastTests` and `StableDiffusionMultiControlNetPipelineFastTests`?

Yes it be great if you could add a test here to move this PR forward @celestialphineas - very exciting PR!

File: src/diffusers/pipelines/controlnet/pipeline_controlnet.py
Comment (yiyixuxu) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet.py:690]: I'm not sure about this design here - I think it is way better we keep this `check_image` method as it is (i.e. only accept PIL, tensor, np, list of pil, list of tensor ad list of np) and handle the the nested image outside this method and inside `check_input`, this way we can:
1. avoid calling `check_image` inside `check_image` - this is definitely not a case where recursion is needed 
2. we can use `check_image` to each element in the nested image

**Review by yiyixuxu — COMMENTED** [review]
Thanks a ton for working on this! 
this is definitely one of the PRs not so straightforward but needed to be done! 

I left feedback about the design, let me know what you think

File: src/diffusers/pipelines/controlnet/pipeline_controlnet.py
Comment (celestialphineas) [reply src/diffusers/pipelines/controlnet/pipeline_controlnet.py:690]: Thank you for the review! Then I guess I should go back to commit 31b531a, and see if there're other ways to refactor.
https://github.com/huggingface/diffusers/pull/6334#discussion_r1436328392

**Comment by celestialphineas** [comment]
> Yes it be great if you could add a test here to move this PR forward @celestialphineas - very exciting PR!

Hi, I'm afraid I'm not very familiar with Diffusers' testing standard. Are there any documentations/PRs to reference? Many thanks.

**Comment by yiyixuxu** [comment]
> > Yes it be great if you could add a test here to move this PR forward @celestialphineas - very exciting PR!
> 
> Hi, I'm afraid I'm not very familiar with Diffusers' testing standard. Are there any documentations/PRs to reference? Many thanks.

take a look at this file - you can just add maybe 2 unit tests there for nested images, one for single controlnet and one for multi
https://github.com/huggingface/diffusers/blob/6ae7e8112a49ac20ade10ac639a76edbd9817a51/tests/pipelines/controlnet/test_controlnet.py#L111

File: src/diffusers/pipelines/controlnet/pipeline_controlnet.py
Comment (celestialphineas) [reply src/diffusers/pipelines/controlnet/pipeline_controlnet.py:690]: https://github.com/huggingface/diffusers/blob/5bacc2f5af7904749827fac15df05df0cb782fe4/src/diffusers/pipelines/controlnet/pipeline_controlnet.py#L732

I believe this line should be `image_batch_size != 1 and image_batch_size != num_images_per_prompt`

Diffusers' current behavior for calling `StableDiffusionControlNetPipeline` looks incorrect when passing a `list` as the `image` argument, even though this was already considered for `StableDiffusionControlNetPipeline` in the code.

https://github.com/huggingface/diffusers/blob/5bacc2f5af7904749827fac15df05df0cb782fe4/src/diffusers/pipelines/controlnet/pipeline_controlnet.py#L611-L612

When `prompt` is a `list` object, it looks as if it should be considered a fixed condition across the `image` input. However, `check_inputs` will simply miss the condition below and throw `RuntimeError` for this:

```python
single_control_pipe(prompt=['cat', 'dog'], image=[img1, img2], num_images_per_prompt=2)
```

```
File ...python3.11/site-packages/torch/utils/_contextlib.py:115, in context_decorator.<locals>.decorate_context(*args, **kwargs)
    112 @functools.wraps(func)
    113 def decorate_context(*args, **kwargs):
    114     with ctx_factory():
...
File ...python3.11/site-packages/torch/nn/modules/linear.py:114, in Linear.forward(self, input)
    113 def forward(self, input: Tensor) -> Tensor:
--> 114     return F.linear(input, self.weight, self.bias)

RuntimeError: mat1 and mat2 must have the same dtype
```

Comment (celestialphineas) [reply src/diffusers/pipelines/controlnet/pipeline_controlnet.py]: I updated with a new commit `50a440f`, where I kept `check_image` unchanged.

Comment (yiyixuxu) [reply src/diffusers/pipelines/controlnet/pipeline_controlnet.py]: @celestialphineas 
Currently, we allow two different behaviors of passing `image` input as a list:
1. `image` should have the same length as `prompt` -  in that case, we will pair each image condition to each prompt 
2. `image` should have the length of `1` - only in this case, we consider it a fixed condition across prompt input

the run time error you received is a `dtype` mismatch error, so I don't think it is related here.

I tested it out too. I think our current implementation works as expected. 

i.e. these two examples work:
`single_control_pipe(prompt=['cat', 'dog'], image=img1, num_images_per_prompt=5)`
`single_control_pipe(prompt=['cat', 'dog'], image=[img1, img2], num_images_per_prompt=2)`

but this one should not 
`single_control_pipe(prompt=['cat', 'dog'], image=[img1, img2, img3], num_images_per_prompt=3)`

this is the code I used 

```python

import torch
from diffusers import ControlNetModel, StableDiffusionControlNetPipeline
from diffusers.utils import load_image



# Loat models
canny_controlnet = ControlNetModel.from_pretrained('lllyasviel/control_v11p_sd15_canny', torch_dtype=torch.float16)

# Single ControlNet pipeline
single_control_pipe = StableDiffusionControlNetPipeline.from_pretrained(
    'runwayml/stable-diffusion-v1-5',
    torch_dtype=torch.float16,
    controlnet=canny_controlnet,
)
single_control_pipe.enable_model_cpu_offload()


canny = load_image("https://huggingface.co/datasets/YiYiXu/testing-images/resolve/main/g_canny.png")

# These images are used for canny control
canny_imgs = [canny] * 3


# Example generation for single_control_pipe
single_controlnet_imgs = single_control_pipe(prompt=['cat', 'dog'], image=[canny] * 2, num_images_per_prompt=2).images
print(len(single_controlnet_imgs))
single_controlnet_imgs = single_control_pipe(prompt=['cat', 'dog'], image=[canny] * 3, num_images_per_prompt=3).images
print(len(single_controlnet_imgs))
```

Comment (celestialphineas) [reply src/diffusers/pipelines/controlnet/pipeline_controlnet.py]: I see. After updating my uneditable Diffusers environment to `0.25.0`, it works as you described. I'm looking into my previous code to see what's happening.

Comment (celestialphineas) [reply src/diffusers/pipelines/controlnet/pipeline_controlnet.py]: Now the multi-ControlNet seems to have the same behavior as its single counterpart, and it will no longer allow input like

```python
double_controlnet_imgs = double_control_pipe(prompt='cat', image=[img1, img2])
```

I still need some further testing, but this PR is becoming simpler than I had expected. :)

File: src/diffusers/pipelines/controlnet/pipeline_controlnet.py
Comment (yiyixuxu) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet.py:619]: can we keep the code change minimum here? I don't think we need to change the code completely - maybe fit the modified logit into the original structure?

Comment (celestialphineas) [reply src/diffusers/pipelines/controlnet/pipeline_controlnet.py]: Hi, sorry for a late reply. Now it seems to have minimal changes compared to the original code.

**Comment by celestialphineas** [comment]
An updated full working example:

```python
import numpy as np
import skimage, torch
from PIL import Image
from diffusers import ControlNetModel, StableDiffusionPipeline, StableDiffusionControlNetPipeline


# Loat models
sd_pipe = StableDiffusionPipeline.from_pretrained('runwayml/stable-diffusion-v1-5', torch_dtype=torch.float16).to('cuda')
canny_controlnet = ControlNetModel.from_pretrained('lllyasviel/control_v11p_sd15_canny/').to('cuda', dtype=sd_pipe.dtype)
tile_controlnet = ControlNetModel.from_pretrained('lllyasviel/control_v11f1e_sd15_tile/').to('cuda', dtype=sd_pipe.dtype)


# Single ControlNet pipeline
single_control_pipe = StableDiffusionControlNetPipeline(
  vae=sd_pipe.vae,
  text_encoder=sd_pipe.text_encoder,
  tokenizer=sd_pipe.tokenizer,
  unet=sd_pipe.unet,
  scheduler=sd_pipe.scheduler,
  safety_checker=sd_pipe.safety_checker,
  feature_extractor=sd_pipe.feature_extractor,
  controlnet=canny_controlnet
)

# Multiple ControlNet pipeline
double_control_pipe = StableDiffusionControlNetPipeline(
  vae=sd_pipe.vae,
  text_encoder=sd_pipe.text_encoder,
  tokenizer=sd_pipe.tokenizer,
  unet=sd_pipe.unet,
  scheduler=sd_pipe.scheduler,
  safety_checker=sd_pipe.safety_checker,
  feature_extractor=sd_pipe.feature_extractor,
  controlnet=[canny_controlnet, tile_controlnet]
)


# Generate test images with sd_pipe
generator_seed = 114514
sd_generated_imgs = sd_pipe(
  prompt='a lovely cat',
  num_inference_steps=20,
  num_images_per_prompt=4,
  generator=torch.Generator(device='cpu').manual_seed(generator_seed)
).images

# These images are used for canny control
canny_imgs = [Image.fromarray(skimage.feature.canny(np.array(img.convert('L')))) for img in sd_generated_imgs]

# These images are used for tile control, they are colored round disks on white backgrounds
def circle_img(color_hex):
  disk_img = np.zeros((512, 512, 3), dtype=np.uint8) + 255
  rr, cc = skimage.draw.disk((256, 256), radius=200)
  # Convert color_hex to RGB array
  disk_img[rr, cc] = np.array([int(color_hex[i:i+2], 16) for i in (1, 3, 5)], dtype=np.uint8)
  return disk_img
tile_imgs = [ 
  Image.fromarray(circle_img(color))
  for color in [ '#89ED87', '#AE45E2', '#E2B845', '#E24545' ]
]


# Example generation for single_control_pipe
single_controlnet_imgs = single_control_pipe(
  prompt=[ 'cat', 'dog' ],
  # This works for fixed image conditioning
  # image=canny_imgs[0],
  # Here, two images correspond to the two prompts respectively
  # image=[ canny_imgs[0], canny_imgs[1] ],
  guidance_scale=4,
  num_inference_steps=20,
  controlnet_conditioning_scale=1.0,
  control_guidance_start=0.0,
  control_guidance_end=0.5,
  guess_mode=True,
  generator=torch.Generator(device='cpu').manual_seed(generator_seed),
  num_images_per_prompt=2
).images

# Example generation for double_control_pipe
double_controlnet_imgs = double_control_pipe(
  prompt=[ 'cat', 'dog' ],
  # This works for fixed image conditioning
  # image=[ canny_imgs[0], tile_imgs[0] ],
  # Now the nested input also works, and two images correspond to the two prompts respectively
  image=[
    [ canny_imgs[0], tile_imgs[0] ],
    [ canny_imgs[1], tile_imgs[1] ]
  ],
  guidance_scale=4,
  num_inference_steps=20,
  controlnet_conditioning_scale=[ 1.0, 1.0 ],
  control_guidance_start=[ 0.0, 0.0 ],
  control_guidance_end=[ 0.4, 0.5 ],
  guess_mode=False,
  generator=torch.Generator(device='cpu').manual_seed(generator_seed),
  num_images_per_prompt=3
).images

skimage.io.imshow(np.concatenate(double_controlnet_imgs, axis=1))
skimage.io.show()
```

File: src/diffusers/pipelines/controlnet/pipeline_controlnet.py
Comment (yiyixuxu) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet.py:638]:
```suggestion
```

I think this will be checked inside `check_image()` - is it the case?  https://github.com/huggingface/diffusers/blob/79df50388df09d9615e3c067695a453bb0a694c0/src/diffusers/pipelines/controlnet/pipeline_controlnet.py#L732

File: src/diffusers/pipelines/controlnet/pipeline_controlnet.py
Comment (yiyixuxu) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet.py:643]:
```suggestion
                        f"For multiple controlnets: if you pass`image` as a list of list, each sublist must have the same length as the number of controlnets, but the sublists in `image` got {len(transposed_image)} images and {len(self.controlnet.nets)} ControlNets."
```

File: tests/pipelines/controlnet/test_controlnet.py
Comment (yiyixuxu) [inline tests/pipelines/controlnet/test_controlnet.py:473]: this is not list of list though, no? it is just a list of 2 images?

**Review by yiyixuxu — COMMENTED** [review]
thanks! left a few more comments
and also I think we can remove this warning now since we now support pass a list of list as image
https://github.com/huggingface/diffusers/blob/79df50388df09d9615e3c067695a453bb0a694c0/src/diffusers/pipelines/controlnet/pipeline_controlnet.py#L609

File: tests/pipelines/controlnet/test_controlnet.py
Comment (celestialphineas) [reply tests/pipelines/controlnet/test_controlnet.py:473]: https://github.com/huggingface/diffusers/blob/79df50388df09d9615e3c067695a453bb0a694c0/tests/pipelines/controlnet/test_controlnet.py#L381-L392

`inputs["image"]` is a list of image, so here comes a list of image list I think.

File: src/diffusers/pipelines/controlnet/pipeline_controlnet.py
Comment (celestialphineas) [reply src/diffusers/pipelines/controlnet/pipeline_controlnet.py:638]: Yes, you are right about this. There's no need for a check out of `check_image()`.

**Comment by celestialphineas** [comment]
I believe this change may also apply to these pipelines:

```
controlnet/pipeline_controlnet_img2img.py
controlnet/pipeline_controlnet_inpaint_sd_xl.py
controlnet/pipeline_controlnet_inpaint.py
controlnet/pipeline_controlnet_sd_xl_img2img.py
controlnet/pipeline_controlnet_sd_xl.py
```
Should I go ahead or start a separate PR?

File: tests/pipelines/controlnet/test_controlnet.py
Comment (yiyixuxu) [inline tests/pipelines/controlnet/test_controlnet.py:481]: let's also test the case where we only have one image vs multiple prompts so we have to replicate image across the prompt input

```suggestion
        assert np.sum(np.abs(image_1 - image_2)) > 1e-3
        
        inputs = self.get_dummy_inputs(device)
        inputs["prompt"] = [inputs["prompt"], inputs["prompt"]]
        output_1 = sd_pipe(**inputs)
        
        assert np.abs(image - output_1.images).max() < 1e-3
```

**Comment by yiyixuxu** [comment]
@celestialphineas I think a separate PR would be easier! but it is up to you :) 
> Should I go ahead or start a separate PR?

File: src/diffusers/pipelines/controlnet/pipeline_controlnet.py
Comment (sayakpaul) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet.py:633]: Super!

File: src/diffusers/pipelines/controlnet/pipeline_controlnet.py
Comment (sayakpaul) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet.py:661]: While we're at it, could we provide a little example in the error message? Something like:

```
"A single batch of varied conditioning scale settings (ex: ...) is not supported at the moment."
```

**Review by sayakpaul — COMMENTED** [review]
Agree with the changes that @yiyixuxu requested for. Apart from those, really nice PR!

**Comment by celestialphineas** [comment]
I'd like to express my appreciation to the Diffusers team for your patient and thorough review! I noticed quite a bunch of files require this change to support nested image conditioning, so I'd better start new PRs for that.

Thank you!

**Review by patrickvonplaten — APPROVED** [review]
PR looks great, good job @celestialphineas ! 

A short entry here: https://huggingface.co/docs/diffusers/using-diffusers/controlnet#multicontrolnet 
(https://github.com/huggingface/diffusers/blob/main/docs/source/en/using-diffusers/controlnet.md#multicontrolnet) would be great

**Review by DN6 — APPROVED** [review]
LGTM 👍🏽 Let's just add a test for MultiControlNet as well.

**Comment by patrickvonplaten** [comment]
> LGTM 👍🏽 Let's just add a test for MultiControlNet as well.

There is already one "fast" test no?

**Comment by celestialphineas** [comment]
> PR looks great, good job @celestialphineas !
> 
> A short entry here: https://huggingface.co/docs/diffusers/using-diffusers/controlnet#multicontrolnet (https://github.com/huggingface/diffusers/blob/main/docs/source/en/using-diffusers/controlnet.md#multicontrolnet) would be great

As the document for MultiControlNet is using `StableDiffusionXLControlNetPipeline` for the example, which hasn't been worked on in this PR, it might be more appropriate to document it in a future PR, I think.

This behavior may also need to be documented somewhere:

https://github.com/huggingface/diffusers/pull/6334#discussion_r1445748776

> Currently, we allow two different behaviors of passing `image` input as a list:
> 
> 1. `image` should have the same length as `prompt` -  in that case, we will pair each image condition to each prompt
> 2. `image` should have the length of `1` - only in this case, we consider it a fixed condition across prompt input

Currently, I suggest mentioning it here:

https://github.com/huggingface/diffusers/blob/2fb404a2d7b758344ed685c934af55b6e68bd798/src/diffusers/pipelines/controlnet/pipeline_controlnet.py#L902-L909

Adding:

```
When `prompt` is a list, and if a list of images is passed for a single ControlNet, each will be paired with each prompt in the `prompt` list. This also applies to multiple ControlNets, where a list of image lists can be passed.
```

**Comment by sayakpaul** [comment]
Sure. But let's be sure to document it because what you worked on is surely gonna benefit the community. So, let's make sure your work is documented :-)

**Comment by DN6** [comment]
> > LGTM 👍🏽 Let's just add a test for MultiControlNet as well.
> 
> There is already one "fast" test no?

Ah whoops, misread. Yeah LGTM 👍🏽

**Comment by sayakpaul** [comment]
Excellent job @celestialphineas!

**Pull request merged**
